sqldelight-core/src/main/kotlin/com/squareup/sqldelight/core/compiler/SelectQueryGenerator.kt

```diff
@@ -60,43 +61,99 @@ class SelectQueryGenerator(val query: NamedQuery) {
-    if (params.isNotEmpty()) {
-      function.addCode(params.joinToCode(prefix = "(", suffix = ")"))
+      // Binds each parameter to the statement:
+      // statement.bindLong(0, id)
+      function.addCode(argument.preparedStatementBinder(index))
     }
 
-    val mapperLamda = CodeBlock.builder().addStatement(" { cursor ->").indent()
+    // Assemble the actual mapper lambda:
+    // { resultSet ->
+    //   mapper(
+    //       resultSet.getLong(0),
+    //       queryWrapper.tableAdapter.columnAdapter.decode(resultSet.getString(0))
+    //   )
+    // }
+    val mapperLamda = CodeBlock.builder().addStatement(" { $RESULT_SET_NAME ->").indent()
 
     if (query.resultColumns.size > 1) {
+      // Function takes a custom mapper.
+
+      // Add the type variable to the signature.
       val typeVariable = TypeVariableName("T")
-      val mapper = ParameterSpec.builder("mapper", LambdaTypeName.get(
+      function.addTypeVariable(typeVariable)
+
+      // Add the custom mapper to the signature:
+      // mapper: (id: kotlin.Long, value: kotlin.String) -> T
+      function.addParameter(ParameterSpec.builder("mapper", LambdaTypeName.get(
           parameters = query.resultColumns.map {
             ParameterSpec.builder(it.name, it.javaType).build()
           },
           returnType = typeVariable
-      )).build()
+      )).build())
+
+      // Specify the return type for the mapper:
+      // Query<T>
       function.returns(ParameterizedTypeName.get(QUERY_TYPE, typeVariable))
-          .addTypeVariable(typeVariable)
-          .addParameter(mapper)
+
+      // Add the call of mapper with the deserialized columns:
+      // mapper(
+      //     resultSet.getLong(0),
+      //     queryWrapper.tableAdapter.columnAdapter.decode(resultSet.getString(0))
+      // )
       mapperLamda.add("mapper(\n")
           .indent()
-      val decoders = query.resultColumns.mapIndexed { index, column -> column.cursorGetter(index) }
+      val decoders = query.resultColumns.mapIndexed { index, column -> column.resultSetGetter(index) }
       mapperLamda.add(decoders.joinToCode(separator = ",\n", suffix = "\n"))
           .unindent()
           .add(")\n")
     } else {
+      // No custom type possible, just returns the single column:
+      // fun selectSomeText(_id): Query<String>
       function.returns(
           ParameterizedTypeName.get(QUERY_TYPE, query.resultColumns.single().javaType))
-      mapperLamda.add(query.resultColumns.single().cursorGetter(0)).add("\n")
+      mapperLamda.add(query.resultColumns.single().resultSetGetter(0)).add("\n")
     }
-    mapperLamda.unindent().add("}")
+    mapperLamda.unindent().add("}\n")
 
-    return function.addCode(mapperLamda.build()).build()
+    if (query.arguments.isEmpty()) {
+      // No need for a custom query type, return an instance of Query:
+      // return Query(statement, selectForId) { resultSet -> ... }
+      return function
+          .addCode(CodeBlock.Builder()
```

Comment:
Do you need this `CodeBlock` (and below)? Couldn't these just be individual calls to `addCode`?